File: software/tests/example_loop.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

HOST = "localhost"
PORT = 4223
UID = "xyz" # Change XYZ to the UID of your DI8 TNG

from tinkerforge.ip_connection import IPConnection, time_data
from tinkerforge.tng_di8 import TNGDI8
import time

import matplotlib.pyplot as plt


if __name__ == "__main__":
    ipcon = IPConnection() # Create IP connection
    di8 = TNGDI8(UID, ipcon) # Create device object

    ipcon.connect(HOST, PORT) # Connect to brickd
    # Don't use device before ipcon is connected

    # Get current value
    data_expected = 0
    t = time.time()
    count = 0

    timestamps = []
    values     = []
    while True:
#    for i in range(50):
        time_data.append((11, time.time()))

#        for i in range(len(time_data)-1):
#            print("{0:2} to {1:2}: {2:f}".format(i, i+1, time_data[i+1][1] - time_data[i][1]))
#        print("\n")
        time_data.clear()


#        time.sleep(0.1)
        timestamp, value = di8.get_value()
#        print(timestamp, value)


        data = 0
        for i in range(len(value)):
            data |= (value[i] << i)

        timestamps.append(timestamp/100)
        values.append(data)

        if data_expected != data:
            print("Unexpected data: {0} vs {1}".format(data_expected, data))

        count += 1
        data_expected = (data + 1) % 256
        new_t = time.time()
        if new_t - t >= 1:
            t = new_t
            print("{0} msg/s".format(count))
            count = 0

    plt.plot(timestamps, values, '.')
    plt.show()
        
    raw_input("Press key to exit\n") # Use input() in Python 3
    ipcon.disconnect()

File: software/tests/c/example_loop.c
#include <stdio.h>
#include <math.h>
#include <inttypes.h>
#include <time.h>

#include "ip_connection.h"
#include "tng_di8.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your TNG DI8

uint64_t get_ms(void) {
    uint64_t ms; // Milliseconds
    time_t  s;  // Seconds
    struct timespec spec;

    clock_gettime(CLOCK_REALTIME, &spec);

    s  = spec.tv_sec;
    ms = spec.tv_nsec / 1.0e6;
    if (ms > 999) {
        s++;
        ms = 0;
    }

	return ms + 1000*s;
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	DI8 di8;
	di8_create(&di8, UID, &ipcon);

	uint32_t count = 0;

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	uint64_t ms = get_ms();
	uint8_t value_expected = 0;
	while(true) {
		// Get current value
		bool value[8];
		uint64_t timestamp;
		if(di8_get_value(&di8, &timestamp, value) < 0) {
			fprintf(stderr, "Could not get value, probably timeout\n");
			return 1;
		}

		uint8_t value_new = 0;
		for(uint8_t i = 0; i < 8; i++) {
			value_new |= value[i] << i;
		}

		if(value_expected != value_new) {
			printf("Unepxected data: %d vs %d\n", value_expected, value_new);
			value_expected = value_new;
		}

		value_expected++;
		count++;

		uint64_t new_ms = get_ms();
		if(new_ms - ms >= 1000) {
			ms = new_ms;
			printf("%d msg/s\n", count);
			count = 0;
		}
	}

	printf("Press key to exit\n");
	getchar();
	di8_destroy(&di8);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}
